src/main/java/ui/components/pickers/LabelPickerDialog.java

```diff
@@ -3,157 +3,210 @@
 import javafx.geometry.Insets;
 import javafx.scene.control.*;
-import javafx.scene.input.KeyCode;
 import javafx.scene.layout.FlowPane;
 import javafx.scene.layout.VBox;
 import javafx.stage.Modality;
 import javafx.stage.Stage;
 
-import java.util.Collections;
+import java.io.IOException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.ResourceBundle;
+import java.util.Set;
 import java.util.stream.Collectors;
 
-public class LabelPickerDialog extends Dialog<List<String>> {
+import org.apache.logging.log4j.Logger;
+
+import ui.UI;
+import util.HTLog;
+import util.Utility;
+
+/**
+ * Serves as a presenter that synchronizes changes in labels with dialog view  
+ * @author jinified
+ *
+ */
+public class LabelPickerDialog extends Dialog<List<String>> implements Initializable {
 
-    private static final int VBOX_SPACING = 105; // seems like some magic number
     private static final int ELEMENT_MAX_WIDTH = 400;
+    private static final Insets GROUPLESS_PAD = new Insets(5, 0, 0, 0);
+    private static final Insets GROUP_PAD = new Insets(0, 0, 10, 10);
+    private static final Logger logger = HTLog.get(LabelPickerDialog.class);
 
     private final LabelPickerUILogic uiLogic;
-    private final TextField textField;
-    private final FlowPane topPane;
-    private final VBox bottomBox;
+    private final List<TurboLabel> repoLabels;
+    private final List<String> initialLabels;
+    private final Set<String> repoLabelsSet;
+    private final TurboIssue issue;
+
+    private List<String> finalAssignedLabels;
+    
+    @FXML
+    private VBox mainLayout;
+    @FXML
+    private Label title;
+    @FXML
+    private FlowPane assignedLabels;
+    @FXML
+    private TextField queryField;
+    @FXML
+    private VBox feedbackLabels;
 
     LabelPickerDialog(TurboIssue issue, List<TurboLabel> repoLabels, Stage stage) {
-        // UI creation
-        initialiseDialog(stage, issue);
-        createButtons();
-        VBox vBox = createVBox();
-        Label titleLabel = createTitleLabel(issue);
-        titleLabel.setTooltip(createTitleTooltip(issue));
-        topPane = createTopPane();
-        textField = createTextField();
-        bottomBox = createBottomBox();
+        this.repoLabels = repoLabels;
+        this.issue = issue;
+        repoLabelsSet = getRepoLabelsSet();
+        initialLabels = issue.getLabels();
+        finalAssignedLabels = new ArrayList<>(initialLabels);
+        uiLogic = new LabelPickerUILogic();
 
+        initUI(stage, issue);
         setupEvents(stage);
-        uiLogic = new LabelPickerUILogic(issue, repoLabels, this);
+        Platform.runLater(queryField::requestFocus);
+    }
 
-        vBox.getChildren().addAll(titleLabel, topPane, textField, bottomBox);
-        getDialogPane().setContent(vBox);
+    void handleLabelClick(PickerLabel label) {
```

Comment:
Can this be made private, and passed as a callback to the PickerLabel instead? Also, maybe position it lower in the class (under the "Event handling" section), so it flows better.